File: app/src/main/java/com/topjohnwu/magisk/utils/RootUtils.java
package com.topjohnwu.magisk.utils;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.topjohnwu.magisk.Config;
import com.topjohnwu.magisk.Const;
import com.topjohnwu.magisk.R;
import com.topjohnwu.superuser.Shell;
import com.topjohnwu.superuser.ShellUtils;
import com.topjohnwu.superuser.io.SuFile;

import java.io.InputStream;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;

public class RootUtils extends Shell.Initializer {

    public static void rmAndLaunch(String rm, ComponentName component) {
        Shell.su(Utils.fmt("(rm_launch %s %s)&", rm, component.flattenToString())).exec();
    }

    public static void reboot() {
        Shell.su("/system/bin/reboot" + (Config.recovery ? " recovery" : "")).submit();
    }

    public static void startActivity(Intent intent) {
        if (intent.getComponent() == null)
            return;
        List<String> args = new ArrayList<>();
        args.add("am");
        args.add("start");
        intentToCommand(intent, args);
        Shell.su(Utils.argsToCommand(args)).exec();
    }

    public static void intentToCommand(Intent intent, List<String> args) {
        if (intent.getAction() != null) {
            args.add("-a");
            args.add(intent.getAction());
        }
        if (intent.getComponent() != null) {
            args.add("-n");
            args.add(intent.getComponent().flattenToString());
        }
        if (intent.getData() != null) {
            args.add("-d");
            args.add(intent.getDataString());
        }
        if (intent.getCategories() != null) {
            for (String cat : intent.getCategories()) {
                args.add("-c");
                args.add(cat);
            }
        }
        if (intent.getType() != null) {
            args.add("-t");
            args.add(intent.getType());
        }
        Bundle extras = intent.getExtras();
        if (extras != null) {
            for (String key : extras.keySet()) {
                Object val = extras.get(key);
                if (val == null)
                    continue;
                Object value = val;
                String arg;
                if (val instanceof String)          arg = "--es";
                else if (val instanceof Boolean)    arg = "--ez";
                else if (val instanceof Integer)    arg = "--ei";
                else if (val instanceof Long)       arg = "--el";
                else if (val instanceof Float)      arg = "--ef";
                else if (val instanceof Uri)        arg = "--eu";
                else if (val instanceof ComponentName) {
                    arg = "--ecn";
                    value = ((ComponentName) val).flattenToString();
                } else if (val instanceof ArrayList) {
                    ArrayList arr = (ArrayList) val;
                    if (arr.size() <= 0)
                    /* Impossible to know the type due to type erasure */
                        continue;

                    if (arr.get(0) instanceof Integer)      arg = "--eial";
                    else if (arr.get(0) instanceof Long)    arg = "--elal";
                    else if (arr.get(0) instanceof Float)   arg = "--efal";
                    else if (arr.get(0) instanceof String)  arg = "--esal";
                    else                                    continue;  /* Unsupported */
                    StringBuilder sb = new StringBuilder();
                    for (Object o : arr) {
                        sb.append(o.toString().replace(",", "\\,"));
                        sb.append(',');
                    }
                    // Remove trailing comma
                    sb.deleteCharAt(sb.length() - 1);
                    value = sb;
                } else if (val.getClass().isArray()) {
                    if (val instanceof int[])           arg = "--eia";
                    else if (val instanceof long[])     arg = "--ela";
                    else if (val instanceof float[])    arg = "--efa";
                    else if (val instanceof String[])   arg = "--esa";
                    else                                continue;  /* Unsupported */
                    StringBuilder sb = new StringBuilder();
                    int len = Array.getLength(val);
                    for (int i = 0; i < len; ++i) {
                        sb.append(Array.get(val, i).toString().replace(",", "\\,"));
                        sb.append(',');
                    }
                    // Remove trailing comma
                    sb.deleteCharAt(sb.length() - 1);
                    value = sb;
                }
                else continue;  /* Unsupported */

                args.add(arg);
                args.add(key);
                args.add(value.toString());
            }
        }
        args.add("-f");
        args.add(String.valueOf(intent.getFlags()));
    }

    @Override
    public boolean onInit(Context context, @NonNull Shell shell) {
        Shell.Job job = shell.newJob();
        if (shell.isRoot()) {
            job.add(context.getResources().openRawResource(R.raw.util_functions))
                .add(context.getResources().openRawResource(R.raw.utils));
            Const.MAGISK_DISABLE_FILE = new SuFile("/cache/.disable_magisk");
            Config.loadMagiskInfo();
        } else {
            InputStream nonroot = context.getResources().openRawResource(R.raw.nonroot_utils);
            job.add(nonroot);
        }

        job.add("mount_partitions", "get_flags", "run_migrations", "export BOOTMODE=true").exec();

        Config.keepVerity = Boolean.parseBoolean(ShellUtils.fastCmd("echo $KEEPVERITY"));
        Config.keepEnc = Boolean.parseBoolean(ShellUtils.fastCmd("echo $KEEPFORCEENCRYPT"));
        Config.recovery = Boolean.parseBoolean(ShellUtils.fastCmd("echo $RECOVERYMODE"));
        return true;
    }
}

File: app/src/main/java/com/topjohnwu/magisk/utils/RootUtils.kt
package com.topjohnwu.magisk.utils

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.topjohnwu.magisk.Const
import com.topjohnwu.magisk.Info
import com.topjohnwu.magisk.R
import com.topjohnwu.magisk.extensions.rawResource
import com.topjohnwu.magisk.extensions.toShellCmd
import com.topjohnwu.superuser.Shell
import com.topjohnwu.superuser.ShellUtils
import com.topjohnwu.superuser.io.SuFile
import java.util.*
import java.lang.reflect.Array as RArray

fun Intent.toCommand(args: MutableList<String>) {
    if (action != null) {
        args.add("-a")
        args.add(action!!)
    }
    if (component != null) {
        args.add("-n")
        args.add(component!!.flattenToString())
    }
    if (data != null) {
        args.add("-d")
        args.add(dataString!!)
    }
    if (categories != null) {
        for (cat in categories) {
            args.add("-c")
            args.add(cat)
        }
    }
    if (type != null) {
        args.add("-t")
        args.add(type!!)
    }
    val extras = extras
    if (extras != null) {
        loop@ for (key in extras.keySet()) {
            val v = extras.get(key) ?: continue
            var value: Any = v
            val arg: String
            when {
                v is String -> arg = "--es"
                v is Boolean -> arg = "--ez"
                v is Int -> arg = "--ei"
                v is Long -> arg = "--el"
                v is Float -> arg = "--ef"
                v is Uri -> arg = "--eu"
                v is ComponentName -> {
                    arg = "--ecn"
                    value = v.flattenToString()
                }
                v is ArrayList<*> -> {
                    if (v.size <= 0)
                    /* Impossible to know the type due to type erasure */
                        continue@loop

                    arg = if (v[0] is Int)
                        "--eial"
                    else if (v[0] is Long)
                        "--elal"
                    else if (v[0] is Float)
                        "--efal"
                    else if (v[0] is String)
                        "--esal"
                    else
                        continue@loop  /* Unsupported */

                    val sb = StringBuilder()
                    for (o in v) {
                        sb.append(o.toString().replace(",", "\\,"))
                        sb.append(',')
                    }
                    // Remove trailing comma
                    sb.deleteCharAt(sb.length - 1)
                    value = sb
                }
                v.javaClass.isArray -> {
                    arg = if (v is IntArray)
                        "--eia"
                    else if (v is LongArray)
                        "--ela"
                    else if (v is FloatArray)
                        "--efa"
                    else if (v is Array<*> && v.isArrayOf<String>())
                        "--esa"
                    else
                        continue@loop  /* Unsupported */

                    val sb = StringBuilder()
                    val len = RArray.getLength(v)
                    for (i in 0 until len) {
                        sb.append(RArray.get(v, i)!!.toString().replace(",", "\\,"))
                        sb.append(',')
                    }
                    // Remove trailing comma
                    sb.deleteCharAt(sb.length - 1)
                    value = sb
                }
                else -> continue@loop
            }  /* Unsupported */

            args.add(arg)
            args.add(key)
            args.add(value.toString())
        }
    }
    args.add("-f")
    args.add(flags.toString())
}

fun startActivity(intent: Intent) {
    if (intent.component == null)
        return
    val args = ArrayList<String>()
    args.add("am")
    args.add("start")
    intent.toCommand(args)
    Shell.su(args.toShellCmd()).exec()
}

class RootUtils : Shell.Initializer() {

    override fun onInit(context: Context?, shell: Shell): Boolean {
        context ?: return false

        val job = shell.newJob()
        if (shell.isRoot) {
            job.add(context.rawResource(R.raw.util_functions))
                    .add(context.rawResource(R.raw.utils))
            Const.MAGISK_DISABLE_FILE = SuFile("/cache/.disable_magisk")
            Info.loadMagiskInfo()
        } else {
            job.add(context.rawResource(R.raw.nonroot_utils))
        }

        job.add("mount_partitions",
                "get_flags",
                "run_migrations",
                "export BOOTMODE=true")
                .exec()

        Info.keepVerity = ShellUtils.fastCmd("echo \$KEEPVERITY").toBoolean()
        Info.keepEnc = ShellUtils.fastCmd("echo \$KEEPFORCEENCRYPT").toBoolean()
        Info.recovery = ShellUtils.fastCmd("echo \$RECOVERYMODE").toBoolean()
        return true
    }

    companion object {

        fun rmAndLaunch(rm: String, component: ComponentName) {
            Shell.su("(rm_launch $rm ${component.flattenToString()})").exec()
        }
    }
}
